DVCodingStyle.md

```diff
@@ -461,6 +462,125 @@ utility macros for even more common code sequences.
+
+1.  Always wrap macro arguments **and** the output as applicable, within
+    parenthesis.
+
+    The following example illustrates how not wrapping the formal macro
+    arguments in parenthesis can cause unexpected behavior.
+    :-1:
+    ```systemverilog
+    `define AND(a, b) a && b
+
+    assign x = `AND(p || q, r) || s;
+    // Gotcha! wrong operator precedence! This results in: p || (q && r) || s which
+    // is different from what was originally intended: ((p || q) && r) || s.
+    ```
+    :+1:
+    ```systemverilog
+    `define AND(a, b) ((a) && (b))
+
+    assign x = `AND(p || q, r) || s;
+    // Correct: ((p || q) && (r)) || s
+    ```
+2.  Macro usage scope and avoiding namespace collisions.
+
+    *   If macros are defined for use in multiple sources, place them in a
+        separate SystemVerilog header file with appropriate `ifdef` guards.
+
+        This header file must not be compilation unit (provided as an input to
+        the tool for compilation). Instead, it should be `` `included`` in the
+        package, module or the interface that intends to invoke them.
+
+        Macro names should be appropriately prefixed with a string to avoid
+        potential namespace collisions. It is typical to use the file name for
+        the prefix string.
+
+        :-1:
+        ```systemverilog
+        `define STRINGIFY(s) `"s`"
+        // Gotcha! UVM library code also defines a macro with the same name.
+        // Depending on the tool, this may be flagged as an error. You may need
+        // this to support non-UVM testbench designs.
+        ```
+        :+1:
+        ```systemverilog
+        // file: dv_macros.h
+        `define DV_STRINGIFY(s) `"s`"
+        ```
+    *   If the scope of the macro usage is limited to the file, `undef` them
+        at the end of the file.
+
+        Macros have a global scope regardless of where they are defined. Some
+        tools do not even warn about namespace collisions, causing unnecessary
+        debug overhead. It is a good practice to `undef` macros at the end of
+        the file so that they are no longer visible to all subsequent
+        compilation units.
+3.  Avoid passing expressions as formal macro arguments.
+
+    As illustrated in the example below, it is better to avoid passing
+    expressions as arguments, since they are replaced verbatim. The macros may
+    reference the input arguments in multiple places, casuing the expressions to
+    be evaluated multiple times, resulting in unforeseen side effects.
+
+    :-1:
+    ```systemverilog
+    `define CHECK_AND_UPDATE(a, b) if (a != b) a = b;
+
+    `CHECK_AND_UPDATE(x, y++)
+    // Gotcha! This results in: if (x != y++) x = y++; x and y now have totally
+    // different and unexpected values.
+    ```
+    :+1:
+    ```systemverilog
+    `define CHECK_AND_UPDATE(a, b) if (a != b) a = b;
+
+    y++;
+    `CHECK_AND_UPDATE(x, y)
+    ```
+4.  Wrap multi-line macros in `begin` and `end`.
```

Comment:
I just re-read the [OpenTitan PR #1918](https://github.com/lowRISC/opentitan/pull/1918) that seems to have prompted this and I don't think this guidance is enough. The problem with the `` `downcast` macro there is that, while it is a single statement, it has an `if` at top-level, so behaves weirdly if followed by an `else`.

This is the problem that C developers usually address with the `do { ... } while (0)` trick. I think the right guidance here is probably something like "Make sure that macros that evaluate to one or more statements behave just like a single statement, wrapping with `begin/end` if necessary". Then give an example of a macro with an `if` statement, together with what can go wrong.

Reply:
Reading this again (prompted by Scott's prod), I think you don't want to talk about multi-line macros at all. The number of lines is not the point; it's the number of statements. Also (as I said above), an `if` statement is especially pesky and needs the begin/end treatment.

If you want a blanket rule, you could say "wrap non-expression macro bodies in `begin/end`", but I'm not keen. I think that it's reasonable to expect a macro author to understand the language, so I prefer what I suggested in the comment above.

Reply:
It turns out that Wilson Snyder has an excellent list of SystemVerilog macro gotchas called [Verilog Preprocessor: Force for \`Good and \`Evil](https://www.veripool.org/papers/Preproc_Good_Evil_SNUGBos10_paper.pdf). On page 5 (under the "Semicolon Eater" heading) he talks about this `if` statement problem and proposes a `do begin ... end while (0)`, just like the usual C solution.

Reply:
I have reworded this now. 
Yup - see the end of the section I am already referencing that same paper :)